tools/depmat.py

```diff
@@ -0,0 +1,181 @@
+		mapping = []
+		for line in file:
+			if line.isspace():
+				break
+			mapping.append(int(line))
+		size = len(mapping)
+		data = list(map(int, INTEGER_PATTERN.findall(file.read())))
+	assert len(data) == 3*size*size
+	return DepMat(size, mapping, data)
+
+def save(depmat, event):
+	"""Saves `.depmat` dump entry to given `event`.
+	
+	The `event` contains the dump name and event hash,
+	but not the `.depmat` suffix. `depmat` is a DepMat object.
+	"""
+	with open_dump_entry(event + '.depmat', 'wt') as file:
+		rowsize = 3 * depmat.size
+		for value in depmat.mapping or range(depmat.size):
+			file.write(str(value) + '\n')
+		for idx, value in enumerate(depmat.data):
+			if idx % 3:
+				file.write(':' + str(value))
+			elif idx % rowsize:
+				file.write(' ' + str(value))
+			else:
+				file.write('\n' + str(value))
+		file.write('\n')
+
+####
+
+def add_to(matrix, addend):
+	assert matrix.mapping is None
+	for addend_y, matrix_y in enumerate(addend.mapping):
+		matrix_row_offset = 3 * matrix.size * matrix_y
+		addend_row_offset = 3 * addend.size * addend_y
+		for addend_x, matrix_x in enumerate(addend.mapping):
+			matrix_cell_offset = matrix_row_offset + 3*matrix_x
+			addend_cell_offset = addend_row_offset + 3*addend_x
+			matrix.data[matrix_cell_offset+0] += addend.data[addend_cell_offset+0]
+			matrix.data[matrix_cell_offset+1] += addend.data[addend_cell_offset+1]
+			matrix.data[matrix_cell_offset+2] += addend.data[addend_cell_offset+2]
+
+def change_mapping(matrix, new_mapping):
+	return DepMat(matrix.size, [new_mapping[old] for old in matrix.mapping], matrix.data)
+
+def empty(size):
+	return DepMat(size, None, [0] * (size*size*3))
+
+####
+
+def html(event):
+	# TODO adapt from `viewers.depmat`
+	print(f'<!doctype html><html><head><title>{event.event_id}.depmat</title></head><body>TODO</body></html>')
+
+def csv(event):
+	# XXX define format
+	print('0,0,0')
+
+def aggregate(events):
+	aggregated_ps, mappings = phasestack.aggregate(events)
+	aggregated_matrix = empty(aggregated_ps.size)
+	for i, event in enumerate(events):
+		matrix = event.get('depmat', load)
+		add_to(aggregated_matrix, change_mapping(matrix, mappings[i]))
+	return aggregated_matrix, aggregated_ps
+
+def diff(from_, to):
+	# TODO do something useful (currently just lame matrix subtraction)
+	from_depmat = from_.get('depmat', load)
+	from_ps = from_.get('phasestack', phasestack.load)
+	neg_from_depmat = DepMat(from_depmat.size, from_depmat.mapping, [-value for value in from_depmat.data])
+	neg_from = Event(depmat=neg_from_depmat, phasestack=from_ps)
+	return aggregate([neg_from, to])
+
+def expand(original_event, new_ps_event):
+	original_matrix = original_event.get('depmat', load)
+	original_ps = original_event.get('phasestack', phasestack.load)
+	new_ps = new_ps_event.get('phasestack', phasestack.load)
+	mapping = phasestack.get_mapping(original_ps, new_ps)
+	new_matrix = change_mapping(original_matrix, mapping)
+	return new_matrix, new_ps
+
+####
```

Comment:
Overall for the methods above, I would make a class DepMat out of the named tuple. Operations like aggregation would then in my opinion nicely map into addition (`__add__`), bulk aggregation could then be implemented by simple calling of sum(). Structuring the code like this could also make it clearer, what happens  with what. Because it would group methods operating on the same thing in one place.